docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+_Pros_
+* Simpler to implement as it does not require any change in the affinity rules upon identification of a zonal failure.
+* Etcd cluster runs with full strength as long as there is a single zone where etcd pods can be rescheduled.
+
+_Cons_
+* It is possible that even when there is no zonal failure, more than one etcd member can be provisioned in a single zone. The chances of that happening are slim as typically there is a dedicate worker pool for hosting etcd pods.
+
+**Variant-#2**
+
+Use `requiredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone` during the initial setup to strictly enforce one etcd member per zone.
+
+If and when a zonal failure is detected then `etcd-druid` should do the following:
+* Remove the PV and PVC for the etcd member in a zone having an outage
+* Change the affinity rules for etcd pods to now use `preferredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone` during the downtime duration of a zone.
+* Delete the etcd pod in the zone which has an outage
+
+This will force the kube-scheduler to schedule the new pod in another zone.
+
+When it is detected that the zone has now recovered then it should re-balance the etcd members. To achieve that the following `etcd-druid` should do the following:
+* Change the affinity rule to again have `requiredDuringSchedulingIgnoredDuringExecution` for `topologyKey: topology.kubernetes.io/zone`
+* Delete an etcd pod from a zone which has 2 pods running. Subsequently also delete the associated PV and PVC.
+
+The kube-scheduler will now schedule this pod in the just recovered zone.
+
+Consequence of doing this is that `etcd-druid`, which today runs with a single replica, now needs to have a HA setup across zones.
+
+_Pros_
+* When all the zones are healthy and available, it ensures that there is at most one pod per zone, thereby providing the desired QoS w.r.t failure tolerance. Only in the case of a zone failure will it relax the rule for spreading etcd members to allowing more than one member in a zone to be provisioned. However this would ideally be temporary.
+* Etcd cluster runs with full strength as long as there is a single zone where etcd pods can be rescheduled.
+
+_Cons_
+* It is complex to implement.
+* Requires `etcd-druid` to be highly available as it now plays a key role in ensuring that the affinity rules are changed and PV/PVC's are deleted.
+
+## Cost Implications on hosting HA control plane
+
+### Compute & Storage
+
+Cost differential as compared to current setup will be due to:
+> Consider a 3 zone cluster (in case of multi-zonal shoot control plane) and 3 node cluster (in case of single-zonal shoot control plane)
+
+* **Machines**: Depending on the number of zones, a minimum of one additional machine per zone will be provisioned.
+* **Persistent Volume**: 2 additional PV's needs to be provisioned for `etcd-main` pods.
```

Comment:
We also need to do it for `etcd-events` since some controllers, esp. in KCM rely on the availability of the events API. In addition, new Kube-Apiserver instances won't start if `etcd-events` is down.

Reply:
Okay, then let's adapt to make it clear to the reader that etcd-events gets 3 replicas as well.

```suggestion
* **Persistent Volume**: 4 additional PVs need to be provisioned for `etcd-main` and `etcd-events` pods.
```